x-pack/plugins/ingest_manager/server/services/epm/archive/index.ts

```diff
@@ -0,0 +1,320 @@
+      `Uploaded archive seems empty. Assumed content type was ${contentType}, check if this matches the archive type.`
+    );
+  }
+  return paths;
+}
+
+// TODO: everything below performs verification of manifest.yml files, and hence duplicates functionality already implemented in the
+// package registry. At some point this should probably be replaced (or enhanced) with verification based on
+// https://github.com/elastic/package-spec/
+
+function parseAndVerifyArchive(paths: string[]): ArchivePackage {
+  // The top-level directory must match pkgName-pkgVersion, and no other top-level files or directories may be present
+  const toplevelDir = paths[0].split('/')[0];
+  paths.forEach((path) => {
+    if (path.split('/')[0] !== toplevelDir) {
+      throw new PackageInvalidArchiveError('Package contains more than one top-level directory.');
+    }
+  });
+
+  // The package must contain a manifest file ...
+  const manifestFile = `${toplevelDir}/manifest.yml`;
+  const manifestBuffer = cacheGet(manifestFile);
+  if (!paths.includes(manifestFile) || !manifestBuffer) {
+    throw new PackageInvalidArchiveError('Package must contain a top-level manifest.yml file.');
+  }
+
+  // ... which must be valid YAML
+  let manifest;
+  try {
+    manifest = yaml.load(manifestBuffer.toString());
+  } catch (error) {
+    throw new PackageInvalidArchiveError(`Could not parse top-level package manifest: ${error}.`);
+  }
+
+  // Package name and version from the manifest must match those from the toplevel directory
+  if (toplevelDir !== `${manifest.name}-${manifest.version}`) {
+    throw new PackageInvalidArchiveError(
+      `Name ${manifest.name} and version ${manifest.version} do not match top-level directory ${toplevelDir}`
+    );
+  }
+
+  const { name, version, description, type, categories, format_version: formatVersion } = manifest;
+  // check for mandatory fields
+  if (!(name && version && description && type && categories && formatVersion)) {
+    throw new PackageInvalidArchiveError(
+      'Invalid top-level package manifest: one or more fields missing of name, version, description, type, categories, format_version'
+    );
+  }
+
+  const datasets = parseAndVerifyDatasets(paths, name, version);
+  const configTemplates = parseAndVerifyConfigTemplates(manifest);
+
+  return {
+    name,
+    version,
+    description,
+    type,
+    categories,
+    format_version: formatVersion,
+    datasets,
+    config_templates: configTemplates,
+  };
+}
+
+function parseAndVerifyDatasets(paths: string[], pkgName: string, pkgVersion: string): Dataset[] {
+  // A dataset is made up of a subdirectory of name-version/dataset/, containing a manifest.yml
+  let datasetPaths: string[] = [];
+  const datasets: Dataset[] = [];
+
+  // pick all paths matching name-version/dataset/DATASET_PATH/...
+  // from those, pick all unique dataset paths
+  paths
+    .filter((path) => path.startsWith(`${pkgName}-${pkgVersion}/dataset/`))
```

Comment:
This will soon be replaced by `data_stream` instead of `dataset`. @jen-huang FYI

Reply:
@ruflin Will the directory `dataset` in the package archive also be renamed to `data_stream`?

Reply:
are the archives automatically bundled by the registry? if so, I think this would be the case